### PR TITLE
[4007] Reactify "Warning: Workflow Cancellation" dialog

### PR DESCRIPTION
**Pull request by jvega190**
https://github.com/craftercms/craftercms/issues/4007

### CODE REVIEW

File: ui/app/src/components/Dialogs/WorkflowCancellationDialog.tsx [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+function WorkflowCancellationContentUI(props: WorkflowCancellationContentUIProps) {
+  const {
+    resource,
+    classes
+  } = props;
+
+  const { workflowAffectedFiles }: { workflowAffectedFiles: WorkflowAffectedItem[] } = resource.read();
+
+  return (
+    // height 254px
```
Comment (rart) [inline ui/app/src/components/Dialogs/WorkflowCancellationDialog.tsx]: x

File: ui/app/src/components/Dialogs/WorkflowCancellationDialog.tsx [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+  })
+);
+
+function WorkflowCancellationContentUI(props: WorkflowCancellationContentUIProps) {
+  const {
+    resource,
+    classes
+  } = props;
+
+  const { workflowAffectedFiles }: { workflowAffectedFiles: WorkflowAffectedItem[] } = resource.read();
```
Comment (rart) [inline ui/app/src/components/Dialogs/WorkflowCancellationDialog.tsx]: Typing here shouldn’t be necessary if the prop/resource is typed correctly

File: ui/app/src/components/Dialogs/WorkflowCancellationDialog.tsx [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+  const {
+    resource,
+    onClose,
+    onDismiss,
+    onContinue,
+    classes
+  } = props;
+
+  const title = 'Warning: Workflow Cancellation';
+  const subtitle = 'Edit will cancel all items that are in the scheduled deployment batch. Please review the list of files below and chose "Continue" to cancel workflow and edit or "Cancel" to remain in your dashboard.';
```
Comment (rart) [inline ui/app/src/components/Dialogs/WorkflowCancellationDialog.tsx]: Remember to translate

File: ui/app/src/components/Dialogs/WorkflowCancellationDialog.tsx [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+      <DialogFooter>
+        <DialogActions>
+          {onClose && (
+            <Button onClick={onClose} variant="outlined">
+              <FormattedMessage id='workflowCancellationCancel' defaultMessage='Cancel' />
+            </Button>
+          )}
+          {onContinue && (
+            <Button onClick={onContinue} variant="contained" color="primary" autoFocus>
+              <FormattedMessage id='workflowCancellationCancel' defaultMessage='Continue' />
```
Comment (rart) [inline ui/app/src/components/Dialogs/WorkflowCancellationDialog.tsx]: use double quotes for jsx

Comment (rart) [reply ui/app/src/components/Dialogs/WorkflowCancellationDialog.tsx]: ids should be scoped - as always

File: ui/app/src/components/Dialogs/WorkflowCancellationDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+  } = props;
+  useUnmount(props.onClosed);
+
+  const siteId = useActiveSiteId();
+  const [workflowAffectedFiles, setWorkflowAffectedFiles] = useState(null);
+  const [error, setError] = useState(null);
+  const authoringBase = useSelection(state => state.env.authoringBase);
+
+  useEffect(() => {
+    getWorkflowAffectedFiles(siteId, path).subscribe(
```
Comment (rart) [inline ui/app/src/components/Dialogs/WorkflowCancellationDialog.tsx]: Dialog to receive items. No fetch should be done in the dialog.

File: ui/app/src/components/Dialogs/WorkflowCancellationDialog.tsx [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+    workflowAffectedFiles,
+    error
+  }), [workflowAffectedFiles, error]);
+
+  const resource = useLogicResource<Return, Source>(workflowCancellationSource, {
+    shouldResolve: (source) => Boolean(source.workflowAffectedFiles),
+    shouldReject: (source) => Boolean(source.error),
+    shouldRenew: (source, resource) => resource.complete,
+    resultSelector: (source) => ({
+      workflowAffectedFiles: source.workflowAffectedFiles
```
Comment (rart) [inline ui/app/src/components/Dialogs/WorkflowCancellationDialog.tsx]: call these items.
but if this is all, why not return directly the items?

File: ui/app/src/models/Item.ts [hunk 7]
```diff
@@ -90,3 +90,8 @@ export interface LegacyItem {
   [prop: string]: any;
 }
 
+export interface WorkflowAffectedItem {
```
Comment (rart) [inline ui/app/src/models/Item.ts]: How is this different from a LegacyItem? Could we just use that?

File: ui/app/src/services/content.ts [hunk 8]
```diff
@@ -1207,6 +1207,13 @@ export function unlock(site: string, path: string): Observable<boolean> {
   );
 }
 
+export function getWorkflowAffectedFiles(site: string, path: string): Observable<WorkflowAffectedItem[]> {
```
Comment (rart) [inline ui/app/src/services/content.ts]: fetchWorkflowAffectedItems

**Review by rart — CHANGES_REQUESTED** [review]

File: static-assets/components/cstudio-common/common-api.js [hunk 9]
```diff
@@ -2533,21 +2533,27 @@ var nodeOpen = false,
...
-                  view.setContent(content);
-                  view.on('continue', function () {
-                    doEdit();
-                  });
+              const eventIdSuccess = 'workflowCancellationDialogContinue';
+              CrafterCMSNext.system.store.dispatch({
+                type: 'SHOW_WORKFLOW_CANCELLATION_DIALOG',
+                payload: {
+                  open: true,
+                  workflowAffectedFiles: content,
```
Comment (rart) [inline static-assets/components/cstudio-common/common-api.js]: 1. workflowAffectedFiles => items
2. parseLegacyItemToSandBoxItem

File: ui/app/src/components/ComponentMenu.tsx [hunk 10]
```diff
@@ -126,21 +128,39 @@ export default function ComponentMenu(props: ComponentMenuProps) {
...
+              type,
+              inProgress: true,
+              showController: !embeddedParentPath && contentTypesBranch.byId?.[item.contentTypeId]?.type === 'page',
+              itemModel: models[modelId],
+              embeddedParentPath
+            };
+
+            if (affectedFiles?.length > 0) {
+              dispatch(showWorkflowCancellationDialog({
+                workflowAffectedFiles: affectedFiles,
```
Comment (rart) [inline ui/app/src/components/ComponentMenu.tsx]: workflowAffectedFiles => items

File: ui/app/src/components/ComponentMenu.tsx [hunk 11]
```diff
@@ -126,21 +128,39 @@ export default function ComponentMenu(props: ComponentMenuProps) {
...
+                onContinue: showEditDialog(editProps)
+              }));
+            } else {
+              dispatch(
+                showEditDialog({
+                  ...editProps
```
Comment (rart) [inline ui/app/src/components/ComponentMenu.tsx]: no spread required

File: ui/app/src/components/Dialogs/WorkflowCancellationDialog.tsx [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+          withEmptyStateProps={{
+            emptyStateProps: {
+              title: (
+                <FormattedMessage
+                  id="publishDialog.noItemsSelected"
+                  defaultMessage="There are no affected files"
+                />
+              )
+            },
+            isEmpty: (value) => value.length === 0
```
Comment (rart) [inline ui/app/src/components/Dialogs/WorkflowCancellationDialog.tsx]: x

File: ui/app/src/models/Item.ts [hunk 13]
```diff
@@ -89,4 +89,3 @@ export interface LegacyItem {
   children: LegacyItem[];
   [prop: string]: any;
 }
-
```
Comment (rart) [inline ui/app/src/models/Item.ts:92]: x

File: ui/app/src/services/content.ts [hunk 14]
```diff
@@ -1207,6 +1207,13 @@ export function unlock(site: string, path: string): Observable<boolean> {
   );
 }
 
+export function fetchWorkflowAffectedFiles(site: string, path: string): Observable<LegacyItem[]> {
```
Comment (rart) [inline ui/app/src/services/content.ts]: 1. fetchWorkflowAffectedFiles => fetchWorkflowAffectedItems
2. parse to sandbox item using parseLegacyToSandbox...

**Review by rart — CHANGES_REQUESTED** [review]

File: static-assets/components/cstudio-common/common-api.js (hunk 9 above)
Comment (jvega190) [reply static-assets/components/cstudio-common/common-api.js]: 1. done
2. The problem is that `parseLegacyItemToSandBoxItem` is expecting certain properties from the legacyItem, and the items from the workflow API only have `browserUri`, `name` and `path`. Should the items be created differently or should the util be updated?

**Review by rart — APPROVED** [review]

**Pull request merged**